### PR TITLE
`src/tox.ini` (coverage:run): Set `concurrency = multiprocessing,threads`

### LINKED ISSUE #36539: Concurrency bugs in codecov (coverage.py) workflow
> ```
> Combined data file .coverage/.coverage.Helios.local.70384.568987
> Skipping duplicate data .coverage/.coverage.Helios.local.71097.566208
> Skipping duplicate data .coverage/.coverage.Helios.local.71074.300096
> ```

We also see these messages in large quantities the Build & Test CI runs, in the section "Prepare coverage results". For example https://github.com/sagemath/sage/actions/runs/6623852735/job/17991733088#step:19:20

> ```
> /Users/kwankyu/GitHub/sage-dev/local/var/lib/sage/venv-python3.10/lib/python3.10/site-packages/coverage/data.py:171: CoverageWarning: Data file '/Users/kwankyu/GitHub/sage-dev/src/.coverage/.coverage.Helios.local.39475.460681' doesn't seem to be a coverage data file: 
>   data._warn(str(exc))
> Couldn't combine data file .coverage/.coverage.Helios.local.39475.460681: Data file '/Users/kwankyu/GitHub/sage-dev/src/.coverage/.coverage.Helios.local.39475.460681' doesn't seem to be a coverage data file: 
> ```

This one also shows up in the CI, for example https://github.com/sagemath/sage/actions/runs/6606391152/job/17942563622#step:19:1162, but with much less frequency. 

This looks to me like a concurrency bug. It may, in fact, help explain the nondeterministic coverage results that we observed even after fixing the random seed. See #35522.

_Originally posted by @mkoeppe in https://github.com/sagemath/sage/issues/36505#issuecomment-1777824058_

### PR DESCRIPTION
**Pull request by mkoeppe**
- Partially addresses #36539

### :memo: Checklist

- [x] The title is concise, informative, and self-explanatory.
- [ ] The description explains in detail what this PR is about.
- [x] I have linked a relevant issue or discussion.
- [ ] I have created tests covering the changes.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by kwankyu** [comment]
Let's run this again.

**Comment by mkoeppe** [comment]
```
Unknown concurrency choices: threads
Error: Process completed with exit code 1.
```
Hm...

File: src/tox.ini [hunk 1]
```diff
@@ -308,7 +308,7 @@ doctest_optionflags = NORMALIZE_WHITESPACE ELLIPSIS
 
 [coverage:run]
 source = sage
-concurrency = multiprocessing
+concurrency = multiprocessing,threads
 data_file = .coverage/.coverage
```
Comment (kwankyu) [inline src/tox.ini:392]: The default is "thread" (without "s").

File: src/tox.ini [hunk 2]
```diff
@@ -308,7 +308,7 @@ doctest_optionflags = NORMALIZE_WHITESPACE ELLIPSIS
 
 [coverage:run]
 source = sage
-concurrency = multiprocessing
+concurrency = multiprocessing,threads
```
Comment (kwankyu) [inline src/tox.ini]:
```suggestion
concurrency = multiprocessing,thread
```

**Comment by mkoeppe** [comment]
Thanks for the fix! Now it seems to work well

**Comment by mkoeppe** [comment]
There are still a few warnings `Couldn't combine data file .coverage/.coverage.406549dc1e48.422.XQfVypyx: Data file '/sage/.coverage/.coverage.406549dc1e48.422.XQfVypyx' doesn't seem to be a coverage data file` though.

**Comment by kwankyu** [comment]
Are we using "--parallel-mode" flag?

**Comment by kwankyu** [comment]
Ah, yes. I see "-p".

**Comment by kwankyu** [comment]
In build.yml, I see
```yml
      - name: Test all files (sage -t --long ${{ matrix.tests }})
        if: (success() || failure()) && steps.container.outcome == 'success'
        run: |
          mkdir .coverage
          rm -rf /sage/.coverage
          ln -s $(pwd)/.coverage /sage/
          cd /sage
          ./sage -python -m pip install coverage
          ./sage -python -m coverage run --rcfile=src/tox.ini src/bin/sage-runtests --force-lib --long -p4 --format github --random-seed=286735480429121101562228604801325644303 ${{ matrix.tests }} 
```
In src/tox.ini, I see
```ini
[testenv:coverage.py]
# https://coverage.readthedocs.io/en/latest/index.html
description =
    run the Sage doctester with Coverage.py
## This toxenv bypasses the virtual environment set up by tox.
passenv = {[sagedirect]passenv}
setenv  = {[sagedirect]setenv}
envdir  = {[sagedirect]envdir}
allowlist_externals = {[sagedirect]allowlist_externals}
commands_pre =
    {env:SAGE} -pip install -U coverage
commands =
    {env:SAGE} --python -m coverage run "{toxinidir}/../venv/bin/sage-runtests" -p 0 {posargs:--all}
commands_post =
    {env:SAGE} --python -m coverage combine
    {env:SAGE} --python -m coverage report   
...
[coverage:run]
source = sage
concurrency = multiprocessing
data_file = .coverage/.coverage
disable_warnings = no-data-collected  
```

There is "-p" in the command in tox.ini, but there is no "-p" in the command in build.yml. Is this flag applied to the coverage command in the workflow?

**Comment by mkoeppe** [comment]
There's a `-p4` in the command

**Comment by kwankyu** [comment]
> There's a `-p4` in the command

isn't it for `src/bin/sage-runtests`? I meant `-p` for `coverage` command...

**Comment by kwankyu** [comment]
> Now it seems to work well

I feel that we did not fix the root problem yet...

**Comment by mkoeppe** [comment]
> > There's a `-p4` in the command
> 
> isn't it for `src/bin/sage-runtests`? I meant `-p` for `coverage` command...

All the `-p` flags are for `sage-runtests`, not for `coverage`

**Comment by kwankyu** [comment]
OK. Then the `-p` flag for `coverage` is
```
  -p, --parallel-mode   Append the machine name, process id and random number
                        to the data file name to simplify collecting data from
                        many processes.
```
Don't we need this flag? (That was perhaps what I was thinking a month ago)

**Comment by mkoeppe** [comment]
> Don't we need this flag?

No, because that is configured in `src/tox.ini` already

**Comment by kwankyu** [comment]
Sorry but where in `src/tox.ini`?

**Comment by mkoeppe** [comment]
It's the very line that is modified here, in the `[coverage:run]` section

**Comment by kwankyu** [comment]
Do you mean ```concurrency = multiprocessing,thread``` implies `--parallel-mode`?

**Comment by kwankyu** [comment]
> > Now it seems to work well
> 
> I feel that we did not fix the root problem yet...

Do you agree with this?

**Comment by kwankyu** [comment]
Anyway, this PR is likely to improve the situation. Let's merge.

**Review by kwankyu — APPROVED** [review]

**Comment by kwankyu** [comment]
Let's keep the issue #36539 open.

**Comment by mkoeppe** [comment]
> > I feel that we did not fix the root problem yet...
> 
> Do you agree with this?

Yes.

Thanks for the review!

**Pull request merged**